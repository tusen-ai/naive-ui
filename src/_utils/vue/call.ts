type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
