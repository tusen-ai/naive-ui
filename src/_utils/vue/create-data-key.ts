export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
