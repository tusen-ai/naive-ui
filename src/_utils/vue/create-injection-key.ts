import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
