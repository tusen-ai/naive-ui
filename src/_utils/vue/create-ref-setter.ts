import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
