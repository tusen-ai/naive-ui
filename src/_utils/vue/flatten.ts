import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
