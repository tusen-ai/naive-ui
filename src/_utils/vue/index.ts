export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
