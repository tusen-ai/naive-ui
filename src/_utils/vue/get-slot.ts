import { type ComponentPublicInstance, type VNodeChild } from 'vue'

export function getSlot (
  instance: ComponentPublicInstance,
  slotName = 'default',
  fallback: VNodeChild[] = []
): VNodeChild[] {
  const slots = instance.$slots
  const slot = slots[slotName]
  if (slot === undefined) return fallback
  return slot()
}
