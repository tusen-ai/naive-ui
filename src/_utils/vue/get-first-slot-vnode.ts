import { type Slots, type VNode } from 'vue'
import { flatten } from './flatten'
import { warn } from '../naive'

export function getFirstSlotVNode (
  slots: Slots,
  slotName = 'default',
  props: unknown = undefined
): VNode | null {
  const slot = slots[slotName]
  if (!slot) {
    warn('getFirstSlotVNode', `slot[${slotName}] is empty`)
    return null
  }
  const slotContent = flatten(slot(props))
  // vue will normalize the slot, so slot must be an array
  if (slotContent.length === 1) {
    return slotContent[0]
  } else {
    warn('getFirstSlotVNode', `slot[${slotName}] should have exactly one child`)
    return null
  }
}
