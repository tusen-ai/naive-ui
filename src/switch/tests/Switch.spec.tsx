import { mount } from '@vue/test-utils'
import { CSSProperties, h } from 'vue'
import { NSwitch } from '../index'

describe('n-switch', () => {
  it('should work with import on demand', () => {
    mount(NSwitch)
  })

  it('should work with `disabled` prop', async () => {
    const wrapper = mount(NSwitch)
    expect(wrapper.find('.n-switch--disabled').exists()).not.toBe(true)

    await wrapper.setProps({ disabled: true })
    expect(wrapper.find('.n-switch').classes()).toContain('n-switch--disabled')
  })

  it('should work with `checked-value` prop', async () => {
    const onUpdateValue = vi.fn()
    const wrapper = mount(NSwitch, {
      props: {
        checkedValue: 'fooo',
        uncheckedValue: 'barr',
        onUpdateValue
      }
    })
    await wrapper.trigger('click')
    expect(onUpdateValue).toHaveBeenCalledWith('fooo')
    await wrapper.trigger('click')
    expect(onUpdateValue).toHaveBeenCalledWith('barr')
    await wrapper.trigger('click')
    expect(onUpdateValue).toHaveBeenCalledWith('fooo')
  })

  it('should work with `checked-value` prop in type layer', () => {
    const onUpdateValue1: (value: string) => void = () => {}
    const onUpdateValue2: (value: number) => void = () => {}
    const onUpdateValue3: (value: boolean) => void = () => {}
    let _ = (
      <NSwitch
        onUpdateValue={onUpdateValue1}
        value={'123'}
        defaultValue={'123'}
      />
    )
    _ = (
      <NSwitch onUpdateValue={onUpdateValue2} value={123} defaultValue={123} />
    )
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _ = (
      <NSwitch
        onUpdateValue={onUpdateValue3}
        value={true}
        defaultValue={false}
      />
    )
  })

  it('should work with `round` prop', async () => {
    const wrapper = mount(NSwitch)
    expect(wrapper.find('.n-switch--round').exists()).toBe(true)
    await wrapper.setProps({ round: false })
    expect(wrapper.find('.n-switch--round').exists()).not.toBe(true)
  })

  it('should work with `size` prop', async () => {
    const wrapper = mount(NSwitch)

    await wrapper.setProps({ size: 'small' })
    expect(wrapper.find('.n-switch').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ size: 'medium' })
    expect(wrapper.find('.n-switch').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ size: 'large' })
    expect(wrapper.find('.n-switch').attributes('style')).toMatchSnapshot()
  })

  it('should work with `value` prop', async () => {
    const wrapper = mount(NSwitch, { props: { value: true } })
    expect(wrapper.find('.n-switch--active').exists()).toBe(true)

    await wrapper.setProps({ value: false })
    expect(wrapper.find('.n-switch--active').exists()).not.toBe(true)
  })

  it('should work with `on-update:value` prop', async () => {
    const onUpdate = vi.fn()
    const wrapper = mount(NSwitch, { props: { 'onUpdate:value': onUpdate } })

    await wrapper.trigger('click')
    expect(onUpdate).toHaveBeenCalled()
  })

  it('should work with `loading` prop', () => {
    const wrapper = mount(NSwitch, {
      props: {
        loading: true
      }
    })
    expect(wrapper.find('.n-base-loading').exists()).toBe(true)
  })

  it('should work with `rail-style` prop', () => {
    const color = 'rgb(32, 128, 240)'
    const railStyle = ({
      focused,
      checked
    }: {
      focused: boolean
      checked: boolean
    }): CSSProperties | string => {
      const style: any = {}
      if (!checked) {
        style.background = color
        if (focused) {
          style.boxShadow = '0 0 0 2px #d0305040'
        }
      }
      return style
    }
    const wrapper = mount(NSwitch, {
      props: {
        railStyle
      }
    })
    expect(wrapper.find('.n-switch__rail').attributes('style')).toContain(color)
  })

  it('should work with slot', () => {
    const wrapper = mount(NSwitch, {
      slots: {
        checked: () => 'checked',
        unchecked: () => 'unchecked'
      }
    })
    expect(wrapper.find('.n-switch__checked').text()).toEqual('checked')
    expect(wrapper.find('.n-switch__unchecked').text()).toEqual('unchecked')
  })
  it('should work with `icon` slot', () => {
    const wrapper = mount(NSwitch, {
      slots: {
        icon: () => h('div', null, 'icon')
      }
    })
    expect(wrapper.find('.n-switch__button').text()).toEqual('icon')
  })
  it('should work with `checked-icon` & `unchecked-icon` slots', async () => {
    const wrapper = mount(NSwitch, {
      slots: {
        'checked-icon': () => h('div', null, 'checked-icon'),
        'unchecked-icon': () => h('div', null, 'unchecked-icon')
      }
    })
    expect(wrapper.find('.n-switch__button').text()).toEqual('unchecked-icon')
    await wrapper.trigger('click')
    expect(wrapper.find('.n-switch__button').text()).toEqual('checked-icon')
  })
})
