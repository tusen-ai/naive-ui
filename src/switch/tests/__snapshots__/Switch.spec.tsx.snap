// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`n-switch should work with \`size\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-button-border-radius: 3px; --n-button-box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.3), inset 0 0 1px 0 rgba(0, 0, 0, 0.05); --n-button-color: #FFF; --n-button-width: 14px; --n-button-width-pressed: 20px; --n-button-height: 14px; --n-height: max(18px, 14px); --n-offset: calc((18px - 14px) / 2); --n-opacity-disabled: 0.5; --n-rail-border-radius: 3px; --n-rail-color: rgba(0, 0, 0, .14); --n-rail-color-active: #18a058; --n-rail-height: 18px; --n-rail-width: 32px; --n-width: max(32px, calc(32px + 14px - 18px)); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-loading-color: #18a058; --n-text-color: white; --n-icon-color: rgb(118, 124, 130);"`;

exports[`n-switch should work with \`size\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-button-border-radius: 3px; --n-button-box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.3), inset 0 0 1px 0 rgba(0, 0, 0, 0.05); --n-button-color: #FFF; --n-button-width: 18px; --n-button-width-pressed: 24px; --n-button-height: 18px; --n-height: max(22px, 18px); --n-offset: calc((22px - 18px) / 2); --n-opacity-disabled: 0.5; --n-rail-border-radius: 3px; --n-rail-color: rgba(0, 0, 0, .14); --n-rail-color-active: #18a058; --n-rail-height: 22px; --n-rail-width: 40px; --n-width: max(40px, calc(40px + 18px - 22px)); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-loading-color: #18a058; --n-text-color: white; --n-icon-color: rgb(118, 124, 130);"`;

exports[`n-switch should work with \`size\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-button-border-radius: 3px; --n-button-box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.3), inset 0 0 1px 0 rgba(0, 0, 0, 0.05); --n-button-color: #FFF; --n-button-width: 22px; --n-button-width-pressed: 28px; --n-button-height: 22px; --n-height: max(26px, 22px); --n-offset: calc((26px - 22px) / 2); --n-opacity-disabled: 0.5; --n-rail-border-radius: 3px; --n-rail-color: rgba(0, 0, 0, .14); --n-rail-color-active: #18a058; --n-rail-height: 26px; --n-rail-width: 48px; --n-width: max(48px, calc(48px + 22px - 26px)); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-loading-color: #18a058; --n-text-color: white; --n-icon-color: rgb(118, 124, 130);"`;
